From 4dd12eaf0a14196f45ce4f25ed4a9f23bb3f55d5 Mon Sep 17 00:00:00 2001
From: Robert Jordens <jordens@gmail.com>
Date: Tue, 8 Jul 2014 22:37:03 -0600
Subject: [PATCH] or1k: initAsmInfo()

---
 lib/Target/OR1K/OR1KTargetMachine.cpp | 1 +
 1 file changed, 1 insertion(+)

diff --git a/lib/Target/OR1K/OR1KTargetMachine.cpp b/lib/Target/OR1K/OR1KTargetMachine.cpp
index 8f260f0..1771bce 100644
--- a/lib/Target/OR1K/OR1KTargetMachine.cpp
+++ b/lib/Target/OR1K/OR1KTargetMachine.cpp
@@ -41,6 +41,7 @@ OR1KTargetMachine(const Target &T, StringRef TT,
              "f64:32:32-v64:32:32-v128:32:32-a0:0:32-n32"),
   InstrInfo(), TLInfo(*this), TSInfo(*this),
   FrameLowering(Subtarget) {
+    initAsmInfo();
 }
 namespace {
 /// OR1K Code Generator Pass Configuration Options.
-- 
1.9.1

