From 7f84a57f4a73837e50df7a8c6905773c0de568ab Mon Sep 17 00:00:00 2001
From: Robert Jordens <jordens@gmail.com>
Date: Tue, 8 Jul 2014 22:40:10 -0600
Subject: [PATCH] use 3.3 opcodes

---
 llvm/core.py | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/llvm/core.py b/llvm/core.py
index 20ad062..d33860f 100644
--- a/llvm/core.py
+++ b/llvm/core.py
@@ -174,7 +174,7 @@ class OpcodeEnum(Enum):
     OPCODE_PTRTOINT       = 42
     OPCODE_INTTOPTR       = 43
     OPCODE_BITCAST        = 44
-    if llvm.version <= (3, 3):
+    if llvm.version <= (3, 3) or True:
         OPCODE_ICMP           = 45
         OPCODE_FCMP           = 46
         OPCODE_PHI            = 47
-- 
1.9.1

